### ISSUE TITLE
[Task 19]  Concurrent Execution for Custom Rewards

### ISSUE THREAD

**Issue by aayc123** [redai-infra/Relax#129]
Task 19

## 1. Repository, Baseline Commit, and Development Branch

- Repository: https://github.com/redai-infra/Relax
- Baseline branch: `main`
- Baseline commit: `039ce876d25540adad847d4223b4de4722d8f425`
- Development branch: `feat/task19-concurrent-custom-rewards`

Development and testing will be based on the commit above. The Pull Request will not contain changes unrelated to Task 19.

## 2. Runtime Environment

| Item         | Configuration                                                |
| ------------ | ------------------------------------------------------------ |
| Image        | `ghcr.io/redai-infra/relaxrl:dev-20260715-8325919e`          |
| Image digest | `sha256:3fa8ce578acda6c829b83016bde42c38fa892681e4f36ca330f545616fe578e2` |
| Python       | 3.12.3                                                       |
| CUDA         | 12.9                                                         |
| PyTorch      | 2.11.0+cu129                                                 |
| Ray          | 2.56.0                                                       |
| SGLang       | 0.5.12.post1                                                 |
| GPU          | 1 × NVIDIA H800 PCIe 80 GB                                   |
| CPU          | 2 × Intel Xeon Platinum 8558, 96 cores / 192 threads         |
| Memory       | 1.0 TiB                                                      |
| OS           | Ubuntu 22.04.5 LTS                                           |

## 3. Launch Script and Complete Commands

This task does not add a training launch script or introduce or modify CLI arguments. The existing `--custom-rm-path`, `--reward-max-concurrency`, and `--reward-num-workers` arguments will remain compatible and control custom reward execution.

Start the test image:

```bash
docker run --rm -it \
  --network host \
  --shm-size=64g \
  -v /data/zn/Relax:/workspace/Relax \
  ghcr.io/redai-infra/relaxrl:dev-20260715-8325919e \
  bash
```

Run targeted Task 19 unit tests:

```bash
cd /workspace/Relax
python -m pytest -v tests/engine/rewards/test_custom_reward_executor.py
```

Run RewardWorker regression tests:

```bash
cd /workspace/Relax
python -m pytest -v \
  tests/engine/rewards/test_reward_worker.py \
  tests/engine/rewards/test_custom_reward_executor.py
```

Run the agentic reward path regression tests:

```bash
cd /workspace/Relax
python -m pytest -v tests/test_agentic_rollout.py
```

Run the reward regression suite:

```bash
cd /workspace/Relax
python -m pytest -v tests/engine/rewards/
```

Run code checks before submission:

```bash
cd /workspace/Relax
pre-commit run --all-files
```

- New parameters: none
- Modified parameters: none

### Related Files

The main files are expected to be:

- `relax/engine/rewards/__init__.py`
- `relax/agentic/pipeline/reward.py`
- `tests/engine/rewards/test_custom_reward_executor.py`
- `tests/engine/rewards/task19_custom_rewards.py`

### Core Functions

The planned execution strategy is:

```text
--custom-rm-path
→ load and cache the custom reward function
→ determine whether the function is synchronous or asynchronous
→ directly await async functions in the caller process
→ dispatch sync functions to the RewardWorker process pool
→ enforce global concurrency through reward_max_concurrency
→ control worker process count through reward_num_workers
→ return the reward or a sample-aware exception
```

The planned changes are:

1. Cache custom reward functions by `custom_rm_path` in `RewardExecutor` to avoid loading them for every sample.
2. Detect whether a custom reward function is synchronous or asynchronous.
3. Preserve async custom reward behavior by directly awaiting it in the caller process without creating a RewardWorker.
4. Dispatch sync custom rewards to the existing RewardWorker pool in independent Ray Actor processes so they do not block the rollout event loop.
5. Cache custom reward functions by path inside each RewardWorker so each function is loaded once per process instead of once per sample.
6. Continue using the global semaphore controlled by `reward_max_concurrency`.
7. Use `reward_num_workers` to control the number of RewardWorker processes available to sync custom rewards.
8. Execute non-group custom reward batches concurrently per sample while preserving input result order.
9. Preserve whole-group custom reward behavior when `group_rm=True`.
10. Include the custom reward path and sample `group_index`, `index`, `session_id`, or batch position when loading or execution fails.
11. Ensure exceptions and cancellations release concurrency permits so later reward requests can still complete.
12. Make agentic reward paths delegate to `relax.engine.rewards`, preventing them from bypassing RewardExecutor and the Worker pool.

### Non-goals

This task will not:

- modify the Task 18 reward registry or format-aware router;
- modify GRPO, PPO, DAPO, or other RL algorithms;
- modify Controller, Ray Serve deployment, or rollout request scheduling;
- modify the public `Sample` data structure;
- add third-party dependencies;
- change existing built-in reward formulas;
- change training dataset formats;
- delete or rename existing public APIs;
- change the `--custom-rm-path` function signature;
- optimize GPU performance, training throughput, or convergence;
- modify CUDA, drivers, or the distributed training environment;
- force async custom rewards into Worker processes.

### Baseline Data and Behavior

At commit `039ce876d25540adad847d4223b4de4722d8f425`:

- `--custom-rm-path` supports loading a custom reward function;
- custom rewards are called in the rollout event-loop process;
- async custom rewards can be directly awaited;
- sync custom rewards do not reuse RewardWorker process isolation or concurrency controls;
- custom reward functions are repeatedly passed through `load_function` in different execution paths;
- `batched_async_rm` passes the complete sample list to a custom reward in one call;
- the agentic reward path bypasses `RewardExecutor` and loads and calls custom rewards directly;
- custom reward exceptions do not consistently include sample location context.

Unit tests will use minimal in-memory `Sample` objects without downloading models or training datasets and without using a GPU. Coverage will include:

- sync custom reward execution in independent processes;
- direct awaiting of async custom rewards;
- the `reward_max_concurrency` concurrency limit;
- the `reward_num_workers` Worker limit;
- custom reward caching and one-time loading;
- per-sample execution and result ordering for non-group batches;
- sync and async group reward compatibility;
- failing-sample identification;
- semaphore release and subsequent execution after an exception;
- unified agentic reward integration.

### Acceptance Metrics

| Requirement               | Measurement                                                  | Target                                                       |
| ------------------------- | ------------------------------------------------------------ | ------------------------------------------------------------ |
| Sync process isolation    | Compare the PID returned by the custom reward with the caller PID | PIDs are different                                           |
| Async compatibility       | Compare the async custom reward PID and check Worker creation | PID is unchanged and no RewardWorker is created              |
| Maximum concurrency       | Calculate peak overlapping intervals from sync reward start/end times | Peak concurrency does not exceed `reward_max_concurrency`     |
| Worker count              | Count unique PIDs returned by enough concurrent tasks        | Process count equals the configured `reward_num_workers`     |
| Function loading          | Record function load counts in the caller and each Worker process | Once per path in each process                                 |
| Per-sample batch execution | Execute sync custom rewards for distinct samples             | Every sample is executed and result order matches input order |
| Group compatibility       | Run sync and async group custom rewards                      | The complete group is called once and returns its reward list |
| Error identification      | Trigger an exception for a specified sample                  | Error includes the path and sample index or batch position   |
| Error release             | With concurrency set to 1, fail one sample and then run a valid sample | The next sample completes within the timeout without deadlock |
| Agentic integration       | Inspect the actual delegate used by agentic helpers          | Delegates to `relax.engine.rewards`                          |
| Reward regression         | `tests/engine/rewards/`                                      | All pass                                                     |
| Agentic regression        | `tests/test_agentic_rollout.py`                              | All pass                                                     |
| Code checks               | `pre-commit run --all-files`                                 | All pass                                                     |

Tests will use deterministic inputs without random sampling. The primary acceptance criterion is that the targeted Task 19 unit tests, reward regression suite, agentic regression tests, and code checks pass completely.

## 6. Expected Delivery Method
Delivery method: GitHub Issue + Pull Request.
